### PR TITLE
Route building architecture

### PR DESCRIPTION
**Pull request by PKWadsy**
Significant refactorings to the route tree builder, as described in #28. Acts as a good base to develop #28, #14 and #4

The `RouteTreeBuilder` has been split into three modules

1. A route tree module that represents a tree of endpoints.
2. A builder which takes a routes folder and builds an endpoint tree
3. An adapter which converts the endpoint tree to an express router

### CODE REVIEW

File: src/core/createKint.ts [hunk 1]
```diff
@@ -11,12 +12,11 @@ export function createExpressKint<GlobalContext>(): KintBuilder<
   Router
 > {
   return {
-    kint: Kint.new<GlobalContext>(),
+    kint: KintEndpointBuilder.new<GlobalContext>(),
```
Comment (joshuabrownenz) [inline src/core/createKint.ts:15]: I like this distinction in naming

File: src/core/endpoint-builder/KintEndpointBuilder.ts [hunk 2]
```diff
@@ -40,16 +41,22 @@ export class Kint<
 
     type Config = {};
 
-    return new Kint<Context, Config, {}>(
+    const handlerBuilder: HandlerBuilder<GlobalContext, Context, Config> = {
+      buildConfigurableHandler: <
+        NewContext extends Context,
+        NewConfig extends Config,
+      >(
+        innerHandler: ConfigurableHandler<NewContext, NewConfig>,
```
Comment (joshuabrownenz) [inline src/core/endpoint-builder/KintEndpointBuilder.ts:49]: Hey sorry, but I'm a bit lost on what this type assertion does for `innerHandler`? Cause you're not actually doing anything to change the `innerHandler`?

File: src/core/endpoint-builder/KintEndpointBuilder.ts [hunk 3]
```diff
@@ -71,62 +78,59 @@ export class Kint<
...
-     * @param innermostHandler The handler that this middleware will wrap.
-     * @returns A new handler which passes the inner handler into it.
-     */
-
     const newHandlerBuilder: HandlerBuilder<
+      GlobalContext,
       Extend<Context, ContextExt, Name>,
       Extend<Config, ConfigExt, Name>
     > = {
+      /**
```
Comment (joshuabrownenz) [inline src/core/endpoint-builder/KintEndpointBuilder.ts:109]: Is it best practice to create a function like this inside another function? Maybe depth of 1 is ok but lets make sure it doesn't get out of hand

File: src/core/endpoint-builder/extractParts.ts [hunk 4]
```diff
@@ -1,37 +1,40 @@
...
 ): {
   config: Config;
   validators: Validators;
-  handler: ConfigurableHandler<WithValid<Context, Validators>, Config>;
+  handler: ConfigurableHandler<
+    WithValid<{ global: GlobalContext }, Validators>,
+    Config
+  >;
 } {
   return {
```
Comment (joshuabrownenz) [inline src/core/endpoint-builder/extractParts.ts:33]: I think you need to add some eslint suppresion here. Unless you can figure out a way to get info info here safely.

File: src/core/endpoint-builder/handlerWithValidators.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
-import { ConfigurableHandler } from "./models/ConfigurableHandler";
-import { WithValid } from "./models/DefineEndpointFunction";
-import { ValidatedData, ValidatorArray } from "./models/Validator";
+import { ConfigurableHandler } from "../models/ConfigurableHandler";
+import { WithValid } from "../models/DefineEndpointFunction";
+import { ValidatedData, ValidatorArray } from "../models/Validator";
 
 /**
```
Comment (joshuabrownenz) [inline src/core/endpoint-builder/handlerWithValidators.ts:5]: es-lint suppresion for the below as assertions are probably a good ideas as well. (You can't avoid as with a reduce function like that)

File: src/core/endpoint-tree/fs-builder/buildEndpointTreeFromDirectory.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  if (methodMatch) {
+    // eslint-disable-next-line no-type-assertion/no-type-assertion -- We know that methodMatch[1] is a Method because of the regex
+    const method = methodMatch[1] as Method;
+
+    const moduleName = path.join(filePathObject.dir, filePathObject.base);
+
+    const module = tryFn(() => require(moduleName));
+
+    if (module instanceof Error) {
+      throw new Error(
```
Comment (joshuabrownenz) [inline src/core/endpoint-tree/fs-builder/buildEndpointTreeFromDirectory.ts:74]: The following two Codecov warnings can easily be addressed by tests. I'm happy to write the tests for you if you want.

File: src/core/endpoint-tree/toExpressRouter.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+) {
+  return function expressHandler(
+    request: ExpressRequest,
+    response: ExpressResponse,
+  ) {
+    const kintRequest: KintRequest = {
+      underlying: request,
+      response: {
+        send(status: number, body: unknown) {
+          throw new KintResponse(body, status);
```
Comment (joshuabrownenz) [inline src/core/endpoint-tree/toExpressRouter.ts:31]: Interesting. It seems like none of the tests are actually sending responses. Probably pretty important to test that functionality works.

File: src/core/endpoint-tree/toExpressRouter.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      },
+    };
+
+    const kintResponse = handler(kintRequest, {
+      global: getContext(),
+    });
+
+    if (kintResponse instanceof Promise) {
+      kintResponse.then((kintResponse) => {
+        response.status(kintResponse.status).send(kintResponse.body);
```
Comment (joshuabrownenz) [inline src/core/endpoint-tree/toExpressRouter.ts:42]: Once again. It's probably worth running a quick test for your handling of promises in as a response.

File: src/core/endpoint-tree/toExpressRouter.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+  for (const ep of endpointTree.endpoints) {
+    const handler = expressHandlerFromKintHandler(ep.handler, getContext);
+
+    const methodAsLower = ep.method.toLowerCase() as Lowercase<Method>;
+
+    expressRouter[methodAsLower]("/", handler);
+  }
+
+  for (const subRoute of endpointTree.subRoutes) {
```
Comment (joshuabrownenz) [inline src/core/endpoint-tree/toExpressRouter.ts:74]: Sorry for going on about tests. A quick test for nested routes would be a good idea.

**Review by joshuabrownenz — CHANGES_REQUESTED** [review]
I'm requesting a couple more tests. I'm happy to write them in another PR if you want me to try get our coverage numbers up a bit. Maybe just check the es-lint suppresion before merging

File: src/core/endpoint-tree/toExpressRouter.ts (hunk 7 above)
Comment (joshuabrownenz) [reply src/core/endpoint-tree/toExpressRouter.ts:31]: i.e. doesn't look like we are actually calling endpoints in our test cases

File: src/core/endpoint-builder/KintEndpointBuilder.ts (hunk 2 above)
Comment (PKWadsy) [reply src/core/endpoint-builder/KintEndpointBuilder.ts:49]: I have spent many hours on this. The logic of the program guarantees that by the time that we run this function (the innermost handler builder), the context object and config object will be built up to match the NewContext and NewConfig objects. I can't seem to find a good way around this. Please let me know if you can think of anything.

File: src/core/endpoint-tree/toExpressRouter.ts (hunk 7 above)
Comment (PKWadsy) [reply src/core/endpoint-tree/toExpressRouter.ts:31]: Yea this is definitely worth testing for sure

File: src/core/endpoint-tree/toExpressRouter.ts (hunk 8 above)
Comment (PKWadsy) [reply src/core/endpoint-tree/toExpressRouter.ts:42]: Agreed

File: src/core/endpoint-tree/toExpressRouter.ts (hunk 9 above)
Comment (PKWadsy) [reply src/core/endpoint-tree/toExpressRouter.ts:74]: Agreed

File: src/core/endpoint-tree/fs-builder/buildEndpointTreeFromDirectory.ts (hunk 6 above)
Comment (PKWadsy) [reply src/core/endpoint-tree/fs-builder/buildEndpointTreeFromDirectory.ts:74]: Yea go for it

File: src/core/endpoint-builder/KintEndpointBuilder.ts (hunk 2 above)
Comment (PKWadsy) [reply src/core/endpoint-builder/KintEndpointBuilder.ts:49]: This is the most confusing typescript I have ever written.

Comment (joshuabrownenz) [reply src/core/endpoint-builder/KintEndpointBuilder.ts:49]: Haha lol. Ok at least you can admit that it's confusing. 

When doing this last set of tests I was struggling to grasp what code wraps what code etc. I wonder if there is a nicer we can start to structure everything so it reads slightly better.

**Comment by joshuabrownenz** [comment]
Closes #7

File: src/core/endpoint-builder/KintEndpointBuilder.ts (hunk 3 above)
Comment (PKWadsy) [reply src/core/endpoint-builder/KintEndpointBuilder.ts:110]: It's not perfect as is. But I think the implementation details can be cleaned up at a later date. I've added a todo for now

File: src/core/endpoint-builder/extractParts.ts (hunk 4 above)
Comment (PKWadsy) [reply src/core/endpoint-builder/extractParts.ts:33]: I don't mind leaving the warnings. It leaves the problem open to other contributors to try to make better

**Comment by joshuabrownenz** [comment]
Ready to merge into develop when you are readu

**Comment by PKWadsy** [comment]
> Ready to merge into develop when you are ready. @PKWadsy

Need a new approval from ya

**Review by joshuabrownenz — APPROVED** [review]
Good to go!

**Pull request merged**